pkg/controllers/node/ipam.go

```diff
@@ -200,110 +548,300 @@ func (c *NodeController) cleanup(calicoNodes map[string][]model.AllocationAttrib
-			vxlan := a.AttrSecondary[ipam.AttributeType] == ipam.AttributeTypeVXLAN
-			wg := a.AttrSecondary[ipam.AttributeType] == ipam.AttributeTypeWireguard
-
-			// Skip any allocations which are not either a Kubernetes pod, or a node's
-			// IPIP, VXLAN or Wireguard address. In practice, we don't expect these, but they might exist.
-			// When they do, they will need to be released outside of this controller in order for
-			// the block to be cleaned up.
-			if (ns == "" || pod == "") && !ipip && !vxlan && !wg {
+			// Set the Kubernetes node field now that we know the kubernetes node name
+			// for this allocation.
+			a.knode = knode
+
+			logc = log.WithFields(a.fields())
+			if a.isWindowsReserved() {
+				// Windows reserved IPs don't need garbage collection. They get released automatically when
+				// the block is released.
+				logc.Debug("Skipping Windows reserved IP address")
+				continue
+			}
+
+			if !a.isPodIP() && !a.isTunnelAddress() {
+				// Skip any allocations which are not either a Kubernetes pod, or a node's
+				// IPIP, VXLAN or Wireguard address. In practice, we don't expect these, but they might exist.
+				// When they do, they will need to be released outside of this controller in order for
+				// the block to be cleaned up.
 				logc.Info("IP allocation on node is from an unknown source. Will be unable to cleanup block until it is removed.")
 				canDelete = false
 				continue
 			}
 
-			// Check to see if the pod still exists. If it does, then we shouldn't clean up
-			// this node, since it might come back online.
-			if knode != "" && pod != "" && c.podExistsOnNode(pod, ns, knode) {
-				logc.WithFields(log.Fields{"pod": pod, "ns": ns}).Debugf("Pod still exists")
+			if a.isTunnelAddress() {
+				// Handle tunnel addresses below.
+				tunnelAddresses = append(tunnelAddresses, a)
+				continue
+			}
+
+			if c.allocationIsValid(a, true) {
+				// Allocation is still valid. We can't cleanup the node yet, even
+				// if it appears to be deleted, because the allocation's validity breaks
+				// our confidence.
 				canDelete = false
-				break
+				a.markValid()
+				continue
+			} else if !kubernetesNodeExists {
+				// The allocation is NOT valid, we can skip the candidacy stage.
+				// We know this with confidence because:
+				// - The node the allocation belongs to no longer exists.
+				// - There pod owning this allocation no longer exists.
+				a.markConfirmedLeak()
+			} else if c.config.LeakGracePeriod != nil {
+				// The allocation is NOT valid, but the Kubernetes node still exists, so our confidence is lower.
+				// Mark as a candidate leak. If this state remains, it will switch
+				// to confirmed after the grace period.
+				a.markLeak(c.config.LeakGracePeriod.Duration)
+			}
+
+			if a.isConfirmedLeak() {
+				// If the address is determined to be a confirmed leak, add it to the index.
+				c.confirmedLeaks[a.handle] = a
+			}
```

Comment:
`else ` delete from `confirmedLeaks` in case the IP has been resurrected since we last looked?